### DISCUSSION TITLE
Power Settings Diag Interface Clarification

### DISCUSSION THREAD

**Discussion by bertoldi-silabs** [openthread/openthread#10611] [General]
Hello, I have some questions about these four diag functions related to power settings:

- `otPlatDiagRadioSetRawPowerSetting`
- `otPlatDiagRadioGetRawPowerSetting`
- `otPlatDiagRadioRawPowerSettingEnable`
- `otPlatDiagRadioGetPowerSettings`

My first question is whether or not these functions make sense to implement in the platform library when `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE` is **disabled**.

The last function in this list, `otPlatDiagRadioGetPowerSettings`, is non-weakly implemented when power calibration is enabled, and it seems like the rest of these functions are appropriate to implement in this case. However, I am unsure if it is sensible for these functions to be implemented if power calibration is disabled. It doesn't appear to me that these are used in this configuration, but I also don't see any basis for this assumption in the diag header or the power calibration code.

My other question is related the power calibration feature with regards to the diag startup process:

 On an SOC device, `Diags::ProcessStart` method calls `otPlatRadioSetTransmitPower`. When `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE` is enabled, power_calibration.cpp also implements `otPlatRadioSetChannelTargetPower`, and the note in this function description suggests that only one of these two otPlatRadio functions should be enabled at one time.

Is `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE` generally supported with SOC configurations or only with RCP configurations? It would be nice if this is clarified, as I'm not sure if there is disagreement with this operation on an SOC, or if this is an unsupported configuration.

I would greatly appreciate any clarity, corrections, or other thoughts on this topic.

**Comment by zhanglongxia** [comment]
<1> These functions could be implemented by the platform when `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE` is disabled. You also can use the default implementation of these functions in [power_calibration.cpp](https://github.com/openthread/openthread/blob/main/src/core/utils/power_calibration.cpp) by enabling the `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE`. The `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE ` just provides you with an option of implementing these functions.

<2> All of these four functions have been declared as [week functions](https://github.com/openthread/openthread/blob/main/src/core/diags/factory_diags.cpp#L1025). Whether to implement these functions depends on your product's requirements. You can implement these functions by yourself, or just use the default implementation in [power_calibration.cpp](https://github.com/openthread/openthread/blob/main/src/core/utils/power_calibration.cpp), or do not implement them if you do not need the power calibration feature.

<3>  For the function `otPlatRadioSetTransmitPower()` and `otPlatRadioSetChannelTargetPower()`, if they are working at the same time, the actual output power of the device will be difficult to determine. For example, the actual output power may don't match the value set by the command `txpower xxx`. So we suggest not enabling them at the same time.

<4> The `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE` should be work in both RCP and SoC mode. But you need to implement a platform module to read power configurations from some place and then call these functions to set the power calibration data in the SoC mode.

**Reply by bertoldi-silabs** [reply]
Thank you @zhanglongxia. I will consider the benefits and product needs when determining whether or not to implement these functions with `OPENTHREAD_CONFIG_POWER_CALIBRATION_ENABLE` disabled.

I don't know if I feel that point 3 has been resolved, as the diag module on an SOC will call `otPlatRadioSetTransmitPower` on startup even if `otPlatRadioSetChannelTargetPower` is enabled, which you suggest is not recommended. My impression is that it may make sense to change the diag code to prevent this, but I wanted to check before planning an update.

**Reply by zhanglongxia** [reply]
I think we may need to remove `otPlatRadioSetTransmitPower()`  on startup of the diag to resolve the conflict.

**Comment by zhanglongxia** [comment]
Refer to the latest update of the platform power related APIs: https://github.com/openthread/openthread/pull/10839
